src/main/java/io/pinecone/PineconeDataPlaneClient.java

```diff
@@ -0,0 +1,222 @@
+                               String namespace,
+                               int topK,
+                               Struct filter,
+                               boolean includeValues,
+                               boolean includeMetadata,
+                               List<Float> vectors,
+                               List<Long> sparseIndices,
+                               List<Float> sparseValues) {
+        QueryRequest.Builder queryRequest = QueryRequest.newBuilder();
+
+        if (id != null && !id.isEmpty() && vectors != null && !vectors.isEmpty()) {
+            throw new PineconeValidationException("Invalid Query Request. Both Vector ID and vectors cannot be present.");
+        }
+
+        if (id != null && !id.isEmpty()) {
+            queryRequest.setId(id);
+        }
+
+        if (namespace != null && !namespace.isEmpty()) {
+            queryRequest.setNamespace(namespace);
+        }
+
+        queryRequest.setTopK(topK)
+                .setIncludeValues(includeValues)
+                .setIncludeMetadata(includeMetadata);
+
+        if (filter != null) {
+            queryRequest.setFilter(filter);
+        }
+
+        if (vectors != null && !vectors.isEmpty()) {
+            queryRequest.addAllVector(vectors);
+        }
+
+        if (sparseIndices != null && sparseValues != null) {
+            if (sparseIndices.size() != sparseValues.size()) {
+                throw new PineconeValidationException("Invalid Upsert Request. Please ensure that both sparse indices and values are of the same length.");
+            }
+
+            queryRequest.setSparseVector(SparseValues.newBuilder()
+                    .addAllIndices(convertUnsigned32IntToSigned32Int(sparseIndices))
+                    .addAllValues(sparseValues)
+                    .build());
+        }
+
+        return blockingStub.query(queryRequest.build());
+    }
+
+    public FetchResponse fetch(List<String> ids,
+                               String namespace) {
+        FetchRequest.Builder fetchRequest = FetchRequest.newBuilder();
+
+        if (ids == null || ids.isEmpty()) {
+            throw new PineconeValidationException("Invalid Fetch Request. Vector IDs must be present");
+        }
+
+        fetchRequest.addAllIds(ids);
+
+        if (namespace != null && !namespace.isEmpty()) {
+            fetchRequest.setNamespace(namespace);
+        }
+
+        return blockingStub.fetch(fetchRequest.build());
+    }
+
+    public UpdateResponse update(String id,
+                                 List<Float> values,
+                                 Struct metadata,
+                                 String namespace,
+                                 List<Long> sparseIndices,
+                                 List<Float> sparseValues) {
+        UpdateRequest.Builder updateRequest = UpdateRequest.newBuilder();
+
+        if (id == null) {
+            throw new PineconeValidationException("Invalid Update Request. Vector ID must be present");
```

Comment:
Not sure how I feel about all the capitalization throughout these errors. I'd lean toward not capitalizing things unless you are referring to a specific Java class or something. I'm not going to clutter this review by  commenting on every occurrence of this, but would be good to handle this consistently across all exceptions.

```suggestion
            throw new PineconeValidationException("Invalid update request. Vector id must be present.");
```

Reply:
No strong opinion on this, will update all of them.